src/tsolvers/egraph/Enode.h

```diff
@@ -59,154 +82,86 @@ struct ELRef {
     inline friend bool operator!= (const ELRef& a1, const ELRef& a2) {return a1.x != a2.x; }
-//    struct ELRef operator() (uint32_t v) { x = v; return *this;}
-//    explicit ELRef(uint32_t v) { x = v; }
 };
 
-// FIXME this is uninitialized right now.
 static struct ELRef ELRef_Undef = {INT32_MAX};
 
-class Extra {
-    struct {
-        ERef        car;
-        ERef        cdr;
-        ERef        next;           // Next node in the class
-        int         size;           // Size of the eq class
-        int         carParentIndex;
-        int         cdrParentIndex;
-    } lst;
-    struct {
-        PTRef       pterm;          // The corresponding pterm
-        ELRef       forbid;         // List of unmergeable Enodes
-        dist_t      dist_classes;   // The bit vector for distinction classes
-        int         dist_index;     // My distinction index
-        // fields related to explanation
-        PtAsgn      exp_reason;
-        ERef        exp_parent;
-        ERef        exp_root;
-        int         exp_time_stamp;
-    } trm;
-    friend class Enode;
-    friend class EnodeAllocator;
-};
-
 class EnodeAllocator;
 
-typedef uint32_t CgId;
-
-class Enode
+class Enode final
 {
 private:
-    // Defined for list and term Enodes
-    void setCar  (ERef er)       { assert(type() != et_symb); ex->lst.car = er; }
-    void setCdr  (ERef er)       { assert(type() != et_symb); ex->lst.cdr = er; }
-    void setPterm(PTRef tr)      { assert(isTerm()); ex->trm.pterm = tr; }
-
-protected:
     static uint32_t cgid_ctr;
 
-    struct {
-        unsigned type      : 2;
-        enodeid_t id       : 30; } header;
-
-    static_assert(sizeof(SymRef) == sizeof(ERef), "Expected size of types does not match");
-    union {
-        SymRef symb;
-        ERef   root;
-    };
-
-    ERef        er;             // Either my eref or reference to the relocated one
-    cgId        cid;            // The congruence id of the enode (defined also for symbols)
-    // Must be last field!
-    Extra       ex[0];          // Enode specific data
+    ERef    root;           // The root of this enode's equivalence class
+    cgId    cid;            // The congruence id of the enode (never changes)
+    ERef    eq_next;           // Next node in this enode's equivalence class
+    int     eq_size;           // Size of this enode's equivalence class
+    PTRef   pterm;          // The corresponding pterm
+    ELRef   forbid;         // List of unmergeable Enodes
+    dist_t  dist_classes;   // The bit vector for distinction classes
+
+    // fields related to explanation
+    PtAsgn      exp_reason;
+    ERef        exp_parent;
+    ERef        exp_root;
+    int         exp_time_stamp;
+
+    // Term representation
+    SymRef symb;
+    uint32_t argSize;
+
+    /*
+     * Pointer to the end of object, where ENodeAllocator allocated enough space for Enode's children and
+     * their use-vector indices.
+     */
+    ERef args[0];
```

Comment:
I wonder if it would make sense to interleave these?  It feels like some of the update algorithms could be written simpler that way, but I didn't try myself.  What do you think?

Reply:
I wanted to say that I would like to keep them separate in order to have an ability to iterate over the children. Then I realized that the only place I iterate over the children is now instead of the removed `EnodeStore::getArgTermsAsVector` method.
So maybe it could be done, I was considering something like `<ERef,UseVectorIndex>` pairs.
But I don't know, it does not look very nice. 
Where do you think the update algorithms could be improved?

I was also thinking about this *zero-lenght* array pattern.
As far as I understood, this is GNU extension, which also happens to be supported by clang. But we would probably have problems if we ever want to support Windows platform as well.
I believe the compilers even warn about this with `-Wpedantic`.
But I have not found a good alternative yet.

Reply:
On the zero-length array situation I wouldn't know...  I don't see the world going towards running smt solvers in windows right now, but obviously a more standard solution would be great.  Let's keep it in mind?

On the interleaving, I'm still not sure (= haven't tried) but it feels like `void Egraph::addToUseVectors(ERef parentRef)` could be implemented nicely using this.

Even if we had to go through all children in an interleaved representation, this could be done somewhat elegantly with the `for (auto [eref, _] : enode)` construct.

Reply:
Anyway we might leave this for another PR if you want and think it's a good idea.

Reply:
Yeah, maybe it is possible.
But let's try it in another PR.